doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Validation that the request comes from a trusted source.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `HTTP REQUESTS` source, and add
+a new endpoint to `environmentd`'s HTTP server `/api/webhook/:id`. When a user creates this new
+kind of source, we'll generate an id, persist the new `Source` in the catalog, use the
+`storage-client` to create a new collection, and finally return a well formatted URL to the user.
+
+When we receive a event we'll use the created `SessionClient` to send the raw `headers` and `body`
+to the coordinator, pack the data into a row, and append it to the necessary collection using a
+new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `HTTP REQUESTS`, for which you can
+specify a `FORMAT`, which denotes how we decode the body of a request, the only supported format
+for the time being is `BYTES`.
+
+```
+CREATE SOURCE <name> FROM HTTP REQUESTS FORMAT BYTES;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type              |
+|---------|-------------------|
+| headers | map[text -> text] |
+| body    | bytea             |
+
+Using `bytea` might not be optimal from a storage perspective (e.g. no chance for MFP Pushdown) but
+it makes this initial attempt at webhooks maximally compatible. On the completion of the statement
+we'll return an ID to the user, which they would specify in the webhook URL. The returned ID would
+be 6 character base62 encoded string, a pattern commonly used in URL shorteners.
+
+> **_Note_**: An alternative to `bytea` encoded bodies would be to use `jsonb`. Looking through a
+few common apps that support webhooks, the payloads all default to JSON, or have JSON as the only
+option. Using `bytea` makes us compatible with all webhooks, with the tradeoff of adding work for
+the user to cast to their desired serialization format downstream.
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:id`. This follows the existing pattern of our other
+two endpoints, whose paths also start with `/api`.
+
+> **_Note_**: The existing Web Sockets API is at `/api/experimental/ws`, we could move the webhooks
```

Comment:
Yeah I like the idea of specifying the webhook source via the name, I updated the doc to reflect that!

In the future I would like to also add a path that uses an "external" id, not the source's `GlobalId`, because I think it gives us more flexibility. Exactly like you mentioned, an ID could prevent your existing tools from finding your new source, or it could work the other way, and you could keep an external id assigned to a source, even if you rename it/move it to a different schema or database.

There is some thinking to do about how exactly this would work and the syntax, I think starting with just the name is a great idea.

Reply:
Starting with just the db/schema/item path sounds great to me!

I have a few concerns about introducing external IDs that I'll note for posterity:

  * External IDs would be a new concept that would have to be explained to users. We have a generally high bar for introducing new concepts, based on the general principle of minimize the number of concepts that users have to learn.
  * The external ID namespace would be shared among all teams. That will pose a scalability problem for larger orgs. The really nice thing about db/schema/item is that it plugs into RBAC. You can rig up RBAC so that e.g. the FOO team only has access to create webhook sources in a database named `foo`. With a single external IDs namespace, teams might fight over external IDs (e.g., `github`), and there'd be no way to enforce that the FOO team had to prefix all of their webhooks with `foo-`, e.g., `foo-github`.
  * There's an alternative design where Materialize assigns the external ID rather than letting users specify it, but then users wouldn't be able to repoint that external ID at a different webhook source.

That said, I do think you have a point that users might want to separate their public webhook API structure from their private database/schema layout. But _that_ said, I think we can get a lot of mileage out of the current design with some documentation that webhook names are part of your public API. We can suggest design patterns like "we recommend that you create _all_ your webhooks in a database named `webhooks`, and think very carefully about the names of the sources you create, because they are a pain to change later!"

So, yeah, sidestepping all those design questions around IDs sound sgreat.

Reply:
Great! I added "consider external IDs to future work" and linked this comment!

Reply:
@benesch one existing SQL concept that can solve your problem is being able to support `INSERT` on a restricted subset of `VIEW` definitions.

Reply:
I'm not sure I understand what problem that solves?

Reply:
@benesch I think this would solve what @andrewrodriguez-m had in mind

> In the future I would like to also add a path that uses an "external" id, not the source's GlobalId, because I think it gives us more flexibility. Exactly like you mentioned, an ID could prevent your existing tools from finding your new source, or it could work the other way, and you could keep an external id assigned to a source, even if you rename it/move it to a different schema or database.

and your concern mentioned above

> External IDs would be a new concept that would have to be explained to users. We have a generally high bar for introducing new concepts, based on the general principle of minimize the number of concepts that users have to learn.

I think with write-through views you can solve the above problem with an existing SQL concept.

Rather than exposing your internal database schema / tables / column names to the public through the webhooks, you would have a dedicated `api` schema where you will define views:

```sql
CREATE api.endpoint AS (
    SELECT 
      col1,
      col2, 
      col3
    FROM
      internal.table
)
```

Writes to `api.endpoint` can be reformulated as writes to the backing `internal.table`. You can also evolve the schema of `internal.table` independently. For example you can decide to rename `col3` into `col4`. Then you will redefine the view as follows:

```sql
CREATE api.endpoint AS (
    SELECT 
      col1,
      col2, 
      col4 as col3
    FROM
      internal.table
)
```

Reply:
I see. Write-through views are a generally interesting concept, but would be a large project in their own right. We found a path forward for webhooks that did not involve external IDs, so I think we're good to kick the can on this discussion down the road.